src/cursor/cur_hs.c

```diff
@@ -199,15 +206,324 @@ __curhs_set_key(WT_CURSOR *cursor, ...)
+
+        /* If the start time point is visible to us, let's return that record. */
+        if (__wt_txn_tw_start_visible(session, &cbt->upd_value->tw))
+            break;
+    }
+
+done:
+err:
+    __wt_scr_free(session, &ds_key);
+    return (ret);
+}
+
+/*
+ * __curhs_search_near --
+ *     WT_CURSOR->search_near method for the hs cursor type.
+ */
+static int
+__curhs_search_near(WT_CURSOR *cursor, int *exactp)
+{
+    WT_CURSOR *file_cursor;
+    WT_CURSOR_HS *hs_cursor;
+    WT_DECL_ITEM(srch_key);
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    int cmp;
+    int exact;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    file_cursor = hs_cursor->file_cursor;
+    *exactp = 0;
+    cmp = 0;
+
+    CURSOR_API_CALL_PREPARE_ALLOWED(cursor, session, search_near, CUR2BT(file_cursor));
+
+    WT_ERR(__wt_scr_alloc(session, 0, &srch_key));
+    /* At least we have the btree id set. */
+    WT_ASSERT(session, F_ISSET(hs_cursor, WT_HSC_BTREE_ID_SET));
+    WT_ERR(__wt_buf_set(session, srch_key, file_cursor->key.data, file_cursor->key.size));
+    WT_ERR_NOTFOUND_OK(__wt_hs_cursor_search_near(session, file_cursor, &exact), true);
+
+    /* Empty history store is fine. */
+    if (ret == WT_NOTFOUND)
+        goto done;
+
+    /*
+     * There are some key fields missing so we are searching a range of keys. Place the cursor at
+     * the start of the range.
+     */
+    if (!F_ISSET(hs_cursor, WT_HSC_COUNTER_SET)) {
+        /*
+         * If we raced with a history store insert, we may be two or more records away from our
+         * target. Keep iterating forwards until we are on or past our target key.
+         *
+         * We can't use the cursor positioning helper that we use for regular reads since that will
+         * place us at the end of a particular key/timestamp range whereas we want to be placed at
+         * the beginning.
+         */
+        if (exact < 0) {
+            while ((ret = __wt_hs_cursor_next(session, file_cursor)) == 0) {
+                WT_ERR(__wt_compare(session, NULL, &file_cursor->key, srch_key, &cmp));
+                if (cmp >= 0)
+                    break;
+            }
+            /* No entries greater than or equal to the key we searched for. */
+            WT_ERR_NOTFOUND_OK(ret, true);
+            if (ret == WT_NOTFOUND)
+                goto done;
+
+            *exactp = cmp == 0 ? 0 : 1;
+        } else
+            *exactp = 1;
+
+        WT_ERR(__curhs_next_visible(session, hs_cursor));
+    }
+    /* Search the closest match that is smaller or equal to the search key. */
+    else {
+        /*
+         * Because of the special visibility rules for the history store, a new key can appear in
+         * between our search and the set of updates that we're interested in. Keep trying until we
+         * find it.
+         *
+         * There may be no history store entries for the given btree id and record key if they have
+         * been removed by WT_CONNECTION::rollback_to_stable.
```

Comment:
I would rather just mention _rollback-to-stable routines_ instead of the actual function name. Function names can change and reference would be incorrect.